### PR TITLE
[weave] Add support for default version override

### PR DESCRIPTION
**Pull request by dntosas**
In this PR, we enable users to override the default version of networking/Weave specs.

### CODE REVIEW

File: pkg/apis/kops/networking.go [hunk 1]
```diff
@@ -86,6 +86,9 @@ type WeaveNetworkingSpec struct {
 	NPCCPULimit *resource.Quantity `json:"npcCPULimit,omitempty"`
 	// NPCExtraArgs are extra arguments that are passed to weave-npc.
 	NPCExtraArgs string `json:"npcExtraArgs,omitempty"`
+
+	// Version overrides the default tag of Weave container image.
+	Version string `json:"version,omitempty"`
```
Comment (hakman) [inline pkg/apis/kops/networking.go:91]: @rifelpet any thoughts if to use Version or ImageName, seems very inconsistent?

```suggestion
	// Version overrides the default tag of Weave container image.
	Version string `json:"version,omitempty"`
```

Comment (dntosas) [reply pkg/apis/kops/networking.go:91]: just for reference, i followed this logic --> https://github.com/kubernetes/kops/blob/27f9114f20f9bd49f10214c7177f474db028c4df/upup/models/cloudup/resources/addons/networking.cilium.io/k8s-1.12.yaml.template#L446

Comment (hakman) [reply pkg/apis/kops/networking.go:91]: No worries, just asking @rifelpet if there's a preferred way to do this.

**Comment by hakman** [comment]
I think you are missing a generated file. Can you check if you missed anything from `make apimachinery` ?

File: pkg/apis/kops/networking.go [hunk 2]
```diff
@@ -86,6 +86,9 @@ type WeaveNetworkingSpec struct {
 	NPCCPULimit *resource.Quantity `json:"npcCPULimit,omitempty"`
 	// NPCExtraArgs are extra arguments that are passed to weave-npc.
 	NPCExtraArgs string `json:"npcExtraArgs,omitempty"`
+
+	// Version overrides the default tag of Weave container image.
```
Comment (johngmyers) [inline pkg/apis/kops/networking.go:90]: I would suggest de-emphasizing the default. Something like "Version specifies the Weave container image tag. The default depends on the kOps version."

**Comment by dntosas** [comment]
> I think you are missing a generated file. Can you check if you missed anything from `make apimachinery` ?



> I think you are missing a generated file. Can you check if you missed anything from `make apimachinery` ?

i have the same problem again, running `make apimachinery` does not generate files. will try to fix my env

File: pkg/apis/kops/networking.go [hunk 3]
```diff
@@ -86,6 +86,9 @@ type WeaveNetworkingSpec struct {
 	NPCCPULimit *resource.Quantity `json:"npcCPULimit,omitempty"`
 	// NPCExtraArgs are extra arguments that are passed to weave-npc.
 	NPCExtraArgs string `json:"npcExtraArgs,omitempty"`
+
```
Comment (hakman) [inline pkg/apis/kops/networking.go:89]:
```suggestion
```

File: pkg/apis/kops/v1alpha2/networking.go [hunk 4]
```diff
@@ -86,6 +86,9 @@ type WeaveNetworkingSpec struct {
 	NPCCPULimit *resource.Quantity `json:"npcCPULimit,omitempty"`
 	// NPCExtraArgs are extra arguments that are passed to weave-npc.
 	NPCExtraArgs string `json:"npcExtraArgs,omitempty"`
+
```
Comment (hakman) [inline pkg/apis/kops/v1alpha2/networking.go:89]:
```suggestion
```

File: pkg/apis/kops/networking.go (hunk 2 above)
Comment (dntosas) [reply pkg/apis/kops/networking.go]: thanks for your suggestion @johngmyers , pushed changes accordingly ^^

**Comment by hakman** [comment]
Looks pretty much ok now.

**Comment by hakman** [comment]
Sorry, was ok before with the extra line. Didn't see that all groups are separated in that struct.

**Comment by dntosas** [comment]
> Looks pretty much ok now.

yes :smiley: ^^

Remember i had the same problem last time and we did the same conversation again, pointing me to put git repo into `GOPATH/k8s.io`.

>Sorry, was ok before with the extra line. Didn't see that all groups are separated in that struct.

Should i add extra line or not?

**Comment by hakman** [comment]
Please add it back.

**Comment by dntosas** [comment]
added some docs as reference ^^

File: docs/networking/weave.md [hunk 5]
```diff
@@ -74,3 +74,13 @@ kops update cluster
 
 Since unencrypted nodes will not be able to connect to nodes configured with encryption enabled, this configuration cannot be changed easily without downtime.
 
+### Override Weave image tag
+
```
Comment (hakman) [inline docs/networking/weave.md:78]:
```suggestion
### Override Weave image tag
{{ kops_feature_table(kops_added_default='1.19', k8s_min='1.12') }}

```

File: docs/networking/weave.md [hunk 6]
```diff
@@ -74,3 +74,13 @@ kops update cluster
 
 Since unencrypted nodes will not be able to connect to nodes configured with encryption enabled, this configuration cannot be changed easily without downtime.
 
+### Override Weave image tag
```
Comment (hakman) [inline docs/networking/weave.md:77]:
```suggestion
### Override Weave image tag
{{ kops_feature_table(kops_added_default='1.19', k8s_min='1.12') }}
```

**Pull request merged**